# HG changeset patch
# User Marco Caliari  <marco.caliari@univr.it>
# Date 1301591999 25200
#      Thu Mar 31 10:19:59 2011 -0700
# Node ID 1f0243949b77c79031f8cd6df8596d8d589e78e0
# Parent  4715ce9f911d15c44236fc020086af7a4a4af277
spdiags.m: Treat empty vector (1x0 or 0x1) the same as diag().

diff --git a/scripts/ChangeLog b/scripts/ChangeLog
--- a/scripts/ChangeLog
+++ b/scripts/ChangeLog
@@ -1,8 +1,12 @@
+2011-03-31  Marco Caliari  <marco.caliari@univr.it>
+
+	* sparse/spdiags.m: Treat empty vector (1x0 or 0x1) the same as diag().
+
 2011-03-31  Rik  <octave@nomad.inbox5.com>
 
 	* statistics/base/cor.m: Increase tolerance of %!tests using random
 	data to guarantee passage.
 
 2011-03-31  Rik  <octave@nomad.inbox5.com>
 
 	* deprecated/module.mk, deprecated/cquad.m: Add deprecated entry for
diff --git a/scripts/sparse/spdiags.m b/scripts/sparse/spdiags.m
--- a/scripts/sparse/spdiags.m
+++ b/scripts/sparse/spdiags.m
@@ -74,15 +74,21 @@ function [A, c] = spdiags (v, c, m, n)
     ## Replace specific diagonals c of m with v,c
     [nr, nc] = size (m);
     B = spdiags (m, c);
     A = m - spdiags (B, c, nr, nc) + spdiags (v, c, nr, nc);
   else
     ## Create new matrix of size mxn using v,c
     [j, i, v] = find (v);
     offset = max (min (c(:), n-m), 0);
-    j = j(:) + offset(i);
+    j = j + offset(i);
     i = j-c(:)(i);
     idx = i > 0 & i <= m & j > 0 & j <= n;
     A = sparse (i(idx), j(idx), v(idx), m, n);
   endif
 
 endfunction
+
+%!test
+%assert(spdiags(zeros(1,0),1,1,1),0)
+
+%!test
+%assert(spdiags(zeros(0,1),1,1,1),0)
